gustaf/create/volumes.py

```diff
@@ -0,0 +1,95 @@
+"""gustaf/create/volumes.py
+
+Routines to create volumes.
+"""
+
+import numpy as np
+import random
+
+from gustaf.volumes import Volumes
+
+
+def extrude_tri_to_tet(tri, thickness = 1., layers = 1, randomize = False):
+    """
+    Given triangular faces, create three-dimensional tetrahedral volumes.
+
+    The direct extrusion of a triangle is a 6-vertex wedge. This is divided into
+    three tetrahedra.
+
+    The method is deterministic by default. It walks over the flat vertices in
+    given order and lifts them up by create volumes on the surrounding faces.
+    The `randomize` argument will walk through the vertices in random order to
+    create a less structured result. The randomization will be repeated for
+    every layer.
+
+    Parameters
+    ----------
+    tri: Faces
+    thickness: float
+    layers: int
+    randomize: bool
+
+    Returns
+    --------
+    tet: Volumes
+    """
+    if layers < 1:
+        raise ValueError("The number of layers must be >1.")
+
+    if tri.get_whatami() != "tri":
+        raise ValueError(
+            "Input to extrude_tri_to_tet needs to be a tri mesh, but it's "
+            + tri.get_whatami()
+        )
+
+    # nodes
+    number_of_2d_nodes = tri.vertices.shape[0]
+    number_of_3d_nodes = (layers + 1) * number_of_2d_nodes
+    # node coordinates
+    vertices_3d = np.zeros([number_of_3d_nodes, 3])
+    vertices_3d[:,:2] = np.tile(tri.vertices, [layers + 1, 1])
+    vertices_3d[:,2] = np.repeat(np.arange(layers + 1) *
+            thickness, number_of_2d_nodes)
+    # elements
+    number_of_2d_faces = tri.faces.shape[0]
+    number_of_3d_volumes = layers * 3 * number_of_2d_faces
+    volumes_3d = np.zeros([number_of_3d_volumes, 4])
+    # this is an array that maps from a 2d vertex index to the current upper
+    # vertex index. we start with identity.
+    top_nodes = np.arange(number_of_2d_nodes)
+
+    # we need a mapping from nodes to elements
+    node_elements = dict()
+    for vertex_id in range(number_of_2d_nodes):
```

Comment:
I will make some minor changes. However, in essence, we would also need to solve a coloring problem here to use vectorization/parallelization.

The `node_elements` dictionary has an irregular structure that is not suitable for numpy, I think, since the number of elements connected to an individual node varies. If we're okay with wasting space the data structure could be implemented as a numpy array that has a lot of `-1` entries.

Reply:
`quick_chat_list.append(this)`